### PR TITLE
add runOptimize call in bitmap write method( resubmit of #52842)

### PR DESCRIPTION
**Pull request by UnamedRus**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

RoaringBitmaps being optimized before serialization.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

```
SELECT length(hex(groupBitmapStateIf(number, NOT ((number % X) = 0)))) / 2 AS res
FROM numbers(1000000)
```

X | Master | PR | PR w/o NOT in cond
-- | -- | -- | --
10 | 131224 | 129816 | 126424
40 | 131224 | 100230 | 50152
50 | 131224 | 80242 | 40152
100 | 131224 | 40242 | 20152
1000 | 131224 | 4241 | 2151
10000 | 131224 | 641 | 351
100000 | 131224 | 281 | 82

Difference from previous version, now we run optimize on top of copy of bitmap.
Unfortunately we need to do that, because runOptimize does inplace "optimization" of bitmap, so it change source data.
So at worst we currently do 2 unnecessary? copies of our bitmap:
1. one to be able run `runOptimize`
2. buffer to which we write serialized version of bitmap

Probably, it's better to move `runOptimize` call to different part of code, like merge function. (in that case, we don't need to do copy of bitmap) as it's actually recommended do that frequently and not do it only once before writing. Because it can keep memory usage down.  (https://github.com/RoaringBitmap/CRoaring/issues/215#issuecomment-573465319)
But, it will not work for single thread scenario (merge is not being run in that case), plus it probably will be too frequent and bring performance slowdown.

May be we can have some heuristic, which will run optimize after X merges and at "finalized" step.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/AggregateFunctions/AggregateFunctionGroupBitmapData.h [hunk 1]
```diff
@@ -151,10 +151,12 @@ class RoaringBitmapWithSmallSet : private boost::noncopyable
         }
         else if (BitmapKind::Bitmap == kind)
         {
-            auto size = roaring_bitmap->getSizeInBytes();
+            std::shared_ptr<RoaringBitmap> bitmap = std::make_shared<RoaringBitmap>(*roaring_bitmap);
```
Comment (evillique) [inline src/AggregateFunctions/AggregateFunctionGroupBitmapData.h:154]: Do we need to make it shared_ptr?

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 62bdf1068af4158391d59885f811e2824f7f58e5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/55044/62bdf1068af4158391d59885f811e2824f7f58e5/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

File: src/AggregateFunctions/AggregateFunctionGroupBitmapData.h (hunk 1 above)
Comment (UnamedRus) [reply src/AggregateFunctions/AggregateFunctionGroupBitmapData.h:154]: I don't think, it's necessary to be shared.

Comment (evillique) [reply src/AggregateFunctions/AggregateFunctionGroupBitmapData.h:154]: If we can make it unique_ptr or even better - allocated on stack it should save us a few allocations.

Comment (UnamedRus) [reply src/AggregateFunctions/AggregateFunctionGroupBitmapData.h:154]: Not sure about stack.
https://github.com/RoaringBitmap/CRoaring/issues/112#issuecomment-385924079

Plus, bitmaps can be huge.

But, i have no objections against unique_ptr

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**